File: problems/ABC147/abc147_e.py
H, W = map(int, input().split())

d = [[0]*(W+1) for _ in range(H+1)]
vmax = (H+W)*80
dp = [[[False]*(vmax+1) for _ in range(W+1)] for _ in range(H+1)]
dp[1][1][0] = True

for i in range(H):
    for j, n in enumerate(map(int, input().split())):
        d[i+1][j+1] = n
for i in range(H):
    for j, n in enumerate(map(int, input().split())):
        d[i+1][j+1] -= n
        d[i+1][j+1] = abs(d[i+1][j+1])

for i in range(1, H+1):
    for j in range(1, W+1):
        v1 = d[i-1][j]
        v2 = d[i][j-1]
        d1 = dp[i-1][j]
        d2 = dp[i][j-1]

        for k in range(v1, vmax+1):
            dp[i][j][k] |= d1[k-v1] 
        for k in range(0, vmax-v1+1):
            dp[i][j][k] |= d1[k+v1]
        for k in range(v2, vmax+1):
            dp[i][j][k] |= d2[k-v2]
        for k in range(0, vmax-v2+1):
            dp[i][j][k] |= d2[k+v2]

v = d[H][W]
ans = vmax
for i,j in enumerate(dp[H][W]):
    if not j:continue
    ans = min(ans, abs(i+v), abs(i-v))

print(ans)

File: problems/ABC147/abc147_e.cpp
#include <bits/stdc++.h>
using namespace std;

int dp[80][80][12801] = {0};
int C[80][80] = {0};

int main(){
    int H, W;
    scanf("%d %d", &H, &W);
    int b;

    for (int i=0;i<H;i++) {
        for (int j=0;j<W;j++) {
            scanf("%d", &b);
            C[i][j] += b;
        }
    }

    for (int i=0;i<H;i++) {
        for (int j=0;j<W;j++) {
            scanf("%d", &b);
            C[i][j] -= b;
            C[i][j] = abs(C[i][j]);
        }
    }

    int vmax = (H+W)*(80);
    dp[0][0][C[0][0]] = 1;

    int k1, k2;
    for (int i=0;i<H;i++) {
        for (int j=0;j<W;j++) {
            for (int k=0;k<vmax+1;k++) {
                if (i+1<H) {
                    k1 = k+C[i+1][j];
                    k2 = abs(k-C[i+1][j]);
                    if (k1<=vmax) {
                        dp[i+1][j][k1] |= dp[i][j][k];
                    }
                    if (k2<=vmax) {
                        dp[i+1][j][k2] |= dp[i][j][k];
                    }
                }
                if (j+1<W) {
                    k1 = k+C[i][j+1];
                    k2 = abs(k-C[i][j+1]);
                    if (k1<=vmax) {
                        dp[i][j+1][k1] |= dp[i][j][k];
                    }
                    if (k2<=vmax) {
                        dp[i][j+1][k2] |= dp[i][j][k];
                    }
                }
            }
        }
    }

    int ans = 0;

    for (int i=0;i<vmax;i++) {
        if(dp[H-1][W-1][i]){
            printf("%d\n", i);
            break;
        }
    }
}